### DISCUSSION TITLE
Order of routes with query params after inline {field} tries to take route params as inline

### DISCUSSION THREAD

**Discussion by ruvenz** [fastapi/fastapi#8559] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import APIRouter, FastAPI
from fastapi.exceptions import HTTPException

class service():

  async def return_value(field):
    return field + " is working"


def generate_router() -> APIRouter:
    router = APIRouter()

    @router.get("/{field}")
    async def in_line(field: str):
        try:
            res = await service.return_value(field)
            if res:
                return res
            else:
                raise HTTPException(404)
        except Exception as e:
            raise e


    @router.get("/test")
    async def query_params(field: str):
        try:
            res = await service.return_value(field)
            if res:
                return res
            else:
                raise HTTPException(404)
        except Exception as e:
            raise e
    

    return router


def main():
    app = FastAPI(title="test", openapi_url="/openapi.json", docs_url="/docs",
                  redoc_url="/redocs")

    app.include_router(generate_router())


    uvicorn.run(app, host='localhost', port=8080)



if __name__ == '__main__':
    main()
```

### Description

The code snippet shows the order of the two routes being -
using the input "123" (as an example):
1. the first one is an inline parameter (using { } ) resulting in the request being `localhost:8080/123`
2. the second one is using query params resulting in the request being `localhost:8080/test?field=123`

if you run this test, you will see that the function that handles the parameter will return accordingly:

1. "123 is working"
2. "test is working"

this is a wrong output. If you set breakpoints you will see that it should write "123 is working". 

if you invert the order of the routes to:

```
@router.get("/test")
    async def query_params(field: str):
        try:
            res = await service.return_value(field)
            if res:
                return res
            else:
                raise HTTPException(404)
        except Exception as e:
            raise e

@router.get("/{field}")
    async def in_line(field: str):
        try:
            res = await service.return_value(field)
            if res:
                return res
            else:
                raise HTTPException(404)
        except Exception as e:
            raise e
```
            
meaning the inline params are at the bottom, you will find this working properly.

### Operating System

macOS

### FastAPI Version

0.68.1

### Python Version

Python 3.9.9

**Comment by hellocoldworld** [comment, marked as answer]
The order of the routes is important in fastapi. This is documented in https://fastapi.tiangolo.com/tutorial/path-params/?h=order#order-matters

**Comment by unidesigner** [comment]
A related question here: When specifying a new route with a decorator, is it possible to make it so that the route is *prepended* and not appended to the list of routers that are checked?

I'm running into this issue when I override routes after some auto-generated routes [using fastapi_crud](https://fastapi-crudrouter.awtkns.com/routing). I want to prepend the override routes.

**Comment by haykkh** [comment]
I would also love to know how to be able to override routes after they have been generated by another class
